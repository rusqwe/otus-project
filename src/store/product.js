import axios from "../mixin/axios.js";
const API_URL_STORE = "https://fakestoreapi.com/",
  API_URL_BIN = "http://httpbin.org/";
const { getData, postData } = axios;

export default {
  state: {
    products: [],
  },
  mutations: {
    SET_PRODUCTS: (state, val) => {
      state.products = val;
    },
    ADD_PRODUCT: (state, val) => {
      state.products.unshift({ ...val, count: 1 });
    },
  },
  actions: {
    getProducts({ commit }) {
      return getData(`${API_URL_STORE}products`, {}).then(
        (res) => {
          const products = res.map((item) => ({ ...item, count: 1 }))
          commit("SET_PRODUCTS", products)
        }
      );
    },
    addProduct({ commit }, product) {
      return postData(`${API_URL_STORE}products`, product).then(
        (res) => {
          commit("ADD_PRODUCT", res)
        }
      );
    }
  },
  getters: {
    products: s => s.products,
  },
}