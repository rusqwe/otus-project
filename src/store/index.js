import { createStore } from 'vuex'
import product from "./product.js";

export default createStore({
  modules: {
    product
  },
  state: {
    carts: [],
    user: {
      login: null,
      lastName: null,
      firstName: null,
      surname: null,
      address: null,
      isAdmin: false
    }
  },
  mutations: {
    ADD_CART: (state, val) => {
      const index = state.carts.findIndex(item => item.id === val.id);
      if (index !== -1) {
        state.carts[index].count++;
      } else {
        state.carts.push({ ...val });
      }
    },
    DELETE_CART: (state, val) => {
      const index = state.carts.findIndex(item => item.id === val.id);
      if (index !== -1) {
        state.carts[index].count = state.carts[index].count - 1;
        if (state.carts[index].count < 1) {
          state.carts.splice(index, 1);
        }
      }
    },
    CLEAR_CARTS: (state) => {
      state.carts = [];
    },
    SET_USER: (state, { login, lastName, firstName, surname, address }) => {
      state.user = {
        login: login || state.user.login,
        lastName: lastName || state.user.lastName,
        firstName: firstName || state.user.firstName,
        surname: surname || state.user.surname,
        address: address || state.user.address,
        isAdmin: (Math.random() < 0.5)
      };
    },
  },
  actions: {
    logout() {
      sessionStorage.removeItem('session');
    }
  },
  getters: {
    carts: s => s.carts,
    user: s => s.user,
  },
})