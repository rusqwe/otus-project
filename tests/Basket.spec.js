import { describe, it, vi } from "vitest";
import { mount } from "@vue/test-utils";
import component from "src/views/Basket.vue";
import store from "src/store/index.js";
describe('Basket view', () => {
  const product1 = {
    "id": 1,
    "title": "",
    "price": 2,
    "description": "",
    "category": "",
    "image": "",
    "count": 0
  };
  const product2 = {
    "id": 2,
    "title": "",
    "price": 4,
    "description": "",
    "category": "",
    "image": "",
    "count": 0
  };
  let wrapper;
  beforeEach(() => {
    wrapper = mount(component, {
      global: {
        provide: {
          store: store
        }
      }
    });
    store.commit('ADD_CART', product1);
    store.commit('ADD_CART', product2);
  })

  it('mount without errors', () => {
    expect(wrapper.exists()).toBeTruthy();
    expect(wrapper.html()).toMatchSnapshot();
  })
  it('render cardProduct', () => {
    const cardProduct = wrapper.find('[data-testid="product-1"]')
    expect(wrapper.exists()).toBeTruthy();
    const counter = cardProduct.find('.label-counter');
    expect(counter.element.innerHTML).toEqual('1');
  })
  it('render buttons on cardProduct', () => {
    const buttons = wrapper.find('[data-testid="product-1"]').findAll('.button-counter');
    expect(buttons).toHaveLength(2);
  })
  it('click on add and delete button on cardProduct', async () => {
    const counter = wrapper.find('[data-testid="product-1"]').find('.label-counter');
    const buttons = wrapper.find('[data-testid="product-1"]').findAll('.button-counter');
    const c0 = Number(counter.element.innerHTML);
    await buttons[1].trigger('click');
    expect(c0 + 1).toBe(Number(counter.element.innerHTML));
    await buttons[0].trigger('click');
    expect(c0).toBe(Number(counter.element.innerHTML));
  })

  it('number of productCards', () => {
    const productCards = wrapper.findAll('.item');
    expect(productCards).toHaveLength(2);
  })

  it('click on clear basket button', async () => {
    const buttonClear = wrapper.find('.button-clear');
    await buttonClear.trigger('click');
    const productCards = wrapper.findAll('.item');
    expect(productCards).toHaveLength(0);
  })

  it('total amount in basket', async () => {
    expect(wrapper.find('.label-total').element.innerHTML).toBe("0.00")
    const buttons1 = wrapper.find('[data-testid="product-1"]').findAll('.button-counter');
    const buttons2 = wrapper.find('[data-testid="product-2"]').findAll('.button-counter');
    await buttons1[1].trigger('click');
    await buttons1[1].trigger('click');
    await buttons2[1].trigger('click');
    expect(wrapper.find('.label-total').element.innerHTML).toBe("8.00")
  })
})