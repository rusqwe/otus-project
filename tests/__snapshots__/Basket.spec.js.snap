// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Basket view > mount without errors 1`] = `
"<div data-v-9fbd96f9=\\"\\"><button data-v-9fbd96f9=\\"\\"> Перейти к оформлению </button>
  <p data-v-9fbd96f9=\\"\\"><label data-v-9fbd96f9=\\"\\">Товаров на сумму:</label> <label data-v-9fbd96f9=\\"\\" class=\\"label-total\\">0.00</label></p><label data-v-9fbd96f9=\\"\\" style=\\"display: none;\\">Форма успешно отправлена</label>
  <div data-v-9fbd96f9=\\"\\" class=\\"item\\">
    <div data-v-d3eb8fbd=\\"\\" data-v-9fbd96f9=\\"\\" class=\\"product-item\\" data-testid=\\"product-1\\">
      <div data-v-d3eb8fbd=\\"\\" class=\\"product-img\\"><img data-v-d3eb8fbd=\\"\\" src=\\"\\" height=\\"115\\"></div>
      <div data-v-d3eb8fbd=\\"\\" class=\\"product-list\\">
        <h3 data-v-d3eb8fbd=\\"\\"></h3><span data-v-d3eb8fbd=\\"\\" title=\\"\\"></span>
      </div><span data-v-d3eb8fbd=\\"\\" class=\\"price\\">0.00$</span><button data-v-9fbd96f9=\\"\\" class=\\"button-counter\\"> - </button><label data-v-9fbd96f9=\\"\\" class=\\"label-counter\\">0</label><button data-v-9fbd96f9=\\"\\" class=\\"button-counter\\"> + </button>
    </div>
  </div>
  <div data-v-9fbd96f9=\\"\\" class=\\"item\\">
    <div data-v-d3eb8fbd=\\"\\" data-v-9fbd96f9=\\"\\" class=\\"product-item\\" data-testid=\\"product-2\\">
      <div data-v-d3eb8fbd=\\"\\" class=\\"product-img\\"><img data-v-d3eb8fbd=\\"\\" src=\\"\\" height=\\"115\\"></div>
      <div data-v-d3eb8fbd=\\"\\" class=\\"product-list\\">
        <h3 data-v-d3eb8fbd=\\"\\"></h3><span data-v-d3eb8fbd=\\"\\" title=\\"\\"></span>
      </div><span data-v-d3eb8fbd=\\"\\" class=\\"price\\">0.00$</span><button data-v-9fbd96f9=\\"\\" class=\\"button-counter\\"> - </button><label data-v-9fbd96f9=\\"\\" class=\\"label-counter\\">0</label><button data-v-9fbd96f9=\\"\\" class=\\"button-counter\\"> + </button>
    </div>
  </div><button data-v-9fbd96f9=\\"\\" class=\\"button-clear\\"> Очистить козину </button>
  <div data-v-3cfe8e2e=\\"\\" data-v-9fbd96f9=\\"\\" id=\\"myModal\\" class=\\"modal\\" style=\\"display: none;\\">
    <div data-v-3cfe8e2e=\\"\\" class=\\"modal-content\\">
      <div data-v-3cfe8e2e=\\"\\" class=\\"header\\"> Заказ </div><span data-v-3cfe8e2e=\\"\\" class=\\"close\\">×</span>
      <p data-v-3cfe8e2e=\\"\\">
      <form data-v-bd32e9b7=\\"\\" data-v-9fbd96f9=\\"\\" novalidate=\\"\\">
        <div data-v-bd32e9b7=\\"\\" class=\\"row\\"><label data-v-bd32e9b7=\\"\\" for=\\"lastName\\">Фамилия:</label><input data-v-bd32e9b7=\\"\\" name=\\"lastName\\">
          <!---->
        </div>
        <div data-v-bd32e9b7=\\"\\" class=\\"row\\"><label data-v-bd32e9b7=\\"\\" for=\\"firstName\\">Имя:</label><input data-v-bd32e9b7=\\"\\" name=\\"firstName\\">
          <!---->
        </div>
        <div data-v-bd32e9b7=\\"\\" class=\\"row\\"><label data-v-bd32e9b7=\\"\\" for=\\"surname\\">Отчество:</label><input data-v-bd32e9b7=\\"\\" type=\\"text\\" name=\\"surname\\"></div>
        <div data-v-bd32e9b7=\\"\\" class=\\"row\\"><label data-v-bd32e9b7=\\"\\" for=\\"address\\">Адрес:</label><input data-v-bd32e9b7=\\"\\" name=\\"address\\">
          <!---->
        </div>
        <div data-v-bd32e9b7=\\"\\" class=\\"row\\"><label data-v-bd32e9b7=\\"\\" for=\\"card\\">Номер карты:</label><input data-v-bd32e9b7=\\"\\" name=\\"card\\">
          <!---->
        </div>
        <div data-v-bd32e9b7=\\"\\" class=\\"row\\"><label data-v-bd32e9b7=\\"\\" for=\\"expiry\\">Срок действия:</label><input data-v-bd32e9b7=\\"\\" name=\\"expiry\\">
          <!---->
        </div><input data-v-bd32e9b7=\\"\\" type=\\"submit\\">
      </form>
      </p>
      <div data-v-3cfe8e2e=\\"\\" class=\\"footer\\"></div>
    </div>
  </div>
</div>"
`;
